File: android/src/main/java/io/agora/rtc/base/Callback.kt
package io.agora.rtc.base

import io.agora.rtc.Constants
import io.agora.rtc.RtcEngine
import kotlin.math.abs

abstract class Callback<T> {
    fun code(code: Int?, data: (() -> T?)? = null) {
        val newCode = code ?: Constants.ERR_NOT_INITIALIZED
        if (newCode == 0) {
            success(data?.let { it() })
        } else {
            failure(newCode.toString(), RtcEngine.getErrorDescription(abs(newCode)))
        }
    }

    abstract fun success(data: T?)

    abstract fun failure(code: String, message: String)
}

File: ios/RCTAgora/Base/Callback.swift
//
//  Callback.swift
//  RCTAgora
//
//  Created by LXH on 2020/4/9.
//  Copyright © 2020 Syan. All rights reserved.
//

import Foundation
import AgoraRtcKit

protocol Callback: class {
    associatedtype T

    func success(_ data: Self.T?)

    func failure(_ code: String, _ message: String)
}

extension Callback {
    func code(_ code: Int32?, _ data: (() -> Self.T?)? = nil) {
        let newCode: Int = Int(code ?? Int32(AgoraErrorCode.notInitialized.rawValue))
        if newCode == 0 {
            success(data?())
        } else {
            failure(String(newCode), AgoraRtcEngineKit.getErrorDescription(abs(newCode)) ?? "")
        }
    }
}
